pkg/services/contexthandler/contexthandler.go

```diff
@@ -206,29 +251,19 @@ func (h *ContextHandler) initContextWithAPIKey(reqContext *models.ReqContext) bo
 	defer span.End()
 
 	// base64 decode key
-	decoded, err := apikeygen.Decode(keyString)
-	if err != nil {
-		reqContext.JsonApiErr(401, InvalidAPIKey, err)
-		return true
-	}
-
-	// fetch key
-	keyQuery := models.GetApiKeyByNameQuery{KeyName: decoded.Name, OrgId: decoded.OrgId}
-	if err := h.SQLStore.GetApiKeyByName(reqContext.Req.Context(), &keyQuery); err != nil {
-		reqContext.JsonApiErr(401, InvalidAPIKey, err)
-		return true
+	// We always need to try old key format first otherwise we'll invalidate all keys
+	// beginning with "gl_"
+	var (
+		apikey *models.ApiKey
+		errKey error
+	)
+	apikey, errKey = h.getAPIKey(reqContext, keyString)
+	if errKey != nil && strings.HasPrefix(keyString, apikeygenprefix.GrafanaPrefix) {
+		apikey, errKey = h.getPrefixedAPIKey(reqContext.Req.Context(), keyString)
 	}
 
-	apikey := keyQuery.Result
-
-	// validate api key
-	isValid, err := apikeygen.IsValid(decoded, apikey.Key)
-	if err != nil {
-		reqContext.JsonApiErr(500, "Validating API key failed", err)
```

Comment:
Did you intend to change the behavior here? It looks as though it will always return 401 even when `apikeygen.IsValid` returns an error. At a glance, it does look like `EncodePassword`, and therefore `apikeygen.IsValid`, wouldn't ever return an error, so this might be a non-issue.

Reply:
On a similar note, it doesn't seem like we would want to return `401` if `h.SQLStore.GetApiKeyByName` or `h.SQLStore.GetAPIKeyByHash` fails due to a transient issue where a retry might result in success.

Reply:
Normally I'm not a fan of returning too much context on where the api token validation failed to the potential attacker trying to auth, but since we already return the error anyway we can use the 500 instead of 401 :+1: